### PR TITLE
Fix hang in VonMises rejection sampling for small values of concentration

### LINKED ISSUE #88443: Hang: sampling VonMises distribution gets stuck in rejection sampling for small kappa
### 🐛 Describe the bug

Sampling the VonMises distribution gets stuck in rejection sampling for small values of the concentration parameter. With the location parameter set to 1, the problem starts around ~1e-4 for single and ~1e-8 for double precision.

```torch.distributions.von_mises.VonMises(torch.Tensor([1]), torch.Tensor([1e-4])).sample()```

Numpy doesn't have this problem, likely because small values of the concentration are caught and handled explicitly, see [here](https://github.com/numpy/numpy/blob/623bc1fae1d47df24e7f1e29321d0c0ba2771ce0/numpy/random/src/legacy/legacy-distributions.c#L421). Implementing this might solve the issue.

### Versions

Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (conda-forge gcc 12.1.0-17) 12.1.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.8.13 (default, Oct 21 2022, 23:50:54)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-52-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA T600 Laptop GPU
Nvidia driver version: 515.76
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.5
[pip3] torch==1.12.1
[pip3] torchaudio==0.12.1
[pip3] torchvision==0.13.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py38h7f8727e_0  
[conda] mkl_fft                   1.3.1            py38hd3c417c_0  
[conda] mkl_random                1.2.2            py38h51133e4_0  
[conda] numpy                     1.21.5           py38h6c91a56_3  
[conda] numpy-base                1.21.5           py38ha15fc14_3  
[conda] pytorch                   1.12.1          py3.8_cuda10.2_cudnn7.6.5_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.12.1               py38_cu102    pytorch
[conda] torchvision               0.13.1               py38_cu102    pytorch

cc @ezyang @gchanan @zou3519 @fritzo @neerajprad @alicanb @nikitaved @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mruberry @rgommers

### PR DESCRIPTION
**Pull request by julian-urban**
Fixes #88443

Forces the internal `dtype` of `torch.distributions.von_mises.VonMises` to be `torch.double` and mirrors the numpy implementation of the second order Taylor expansion for `concentration < 1e-5`. Samples and log probs are returned with `dtype` of argument `loc`.

In principle one could also use masking in the rejection sampler to return uniformly distributed numbers for `concentration < 1e-8`, as in numpy. This may be slightly more efficient, but isn't required to solve the hanging issue.

cc @fritzo @neerajprad @alicanb @nikitaved

### CODE REVIEW

File: torch/distributions/von_mises.py [hunk 1]
```diff
@@ -124,14 +124,19 @@ class VonMises(Distribution):
     has_rsample = False
 
     def __init__(self, loc, concentration, validate_args=None):
-        self.loc, self.concentration = broadcast_all(loc, concentration)
+        self.dtype = loc.dtype
+        self.loc, self.concentration = broadcast_all(loc.to(torch.double), concentration.to(torch.double))
```
Comment (fritzo) [inline torch/distributions/von_mises.py:128]: It seems a safer to use double versions only in `.sample()`, maybe storing a
```
self._concentration = self.concentration.to(torch.double)
```
for use in `self._proposal_r`.  This way `.log_prob()` can remain single precision, and `.mean` won't magically cast to double.

If we're going to use double stuff in `_proposal_r`, we should change it to a lazy property:
```python
@lazy_property
def _loc(self):
    return self.loc.to(torch.double)

@lazy_property
def _concentration(self):
    return self.concentration.to(torch.double)

@lazy_property
def _proposal_r(self):
    if self._concentration < 1e-5:
        # second order Taylor expansion around kappa = 0
        self._proposal_r = 1 / self._concentration + self._concentration
    else:
        tau = 1 + (1 + 4 * self._concentration**2).sqrt()
        rho = (tau - (2 * tau).sqrt()) / (2 * self._concentration)
        self._proposal_r = (1 + rho**2) / (2 * rho)
```
Then we can use `self._loc` and `self._concentration` in `.sample()`, but continue using `.loc` and `concentration` elsewhere.

**Review by fritzo — CHANGES_REQUESTED** [review]
Thanks for fixing this!  Could we move your logic to three `@lazy_property`s, to keep the other code paths entirely single precision?

File: torch/distributions/von_mises.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions/von_mises.py:128]: Actually, concentration can be a tensor, so we'll need a `torch.where(self._concentration < 1e-5, -, -)`

**Comment by julian-urban** [comment]
@fritzo Thanks for picking it up, please check my latest commit. I followed your suggestion of implementing the `_loc` and `_concentration` attributes with double prec as lazy properties and similarly `_proposal_r`, also replacing the `if/else` with `torch.where` to make sure it works with tensor-valued `loc` and `concentration`. The hard-coded double prec stuff should now be confined to calls of `sample()` only.

I'm not sure if I was supposed to resolve the above conversation already, this is my first PR and I'm not familiar with the proper etiquette. Apologies.

**Review by fritzo — APPROVED** [review]
Looks great @julian-urban!  It might be worth adding a line comment in `.sample()` explaining why sampling is performed using double precision, but feel free to leave as is.

**Comment by julian-urban** [comment]
@fritzo whoopsie. The test passes now and I added a clarifying comment to `sample()`.

**Review by fritzo — APPROVED** [review]

**Comment by fritzo** [comment]
@ezyang could you please merge this?
